### ISSUE TITLE
Add hash table of game items to game struct and refactor/reimplement Game-State code that relates to finding items when they were stored in a hashtable within the room struct

### ISSUE THREAD

**Issue by patrickkoopmans** [uchicago-cs/chiventure#331]
Due to recent developments in issue #316 and #315 the WDL and game-state team find it necessary to create an item hash table within the game struct which will store pointers to call items in the game using the item_id as a key. This makes several functions such as get_item_from_room() in game.h void, as we no longer need the room id in order to find this item. This issue should only be marked complete when every single redundant function has been deleted--it is more of a searching task than anything. 

It should be noted that WDL has previously specified that all items in a game need to have different IDs, and that this specification was not borne out of a need to hash these items all in one table.

**Comment by jessxia** [comment]
Added hash table of items (all_items) into game struct. 

Added functions into game.h/game.c:
int add_item_to_game(game_t *game, item_t *item);
item_t *get_item_from_game(game_t *g, char *item_id);

Things to consider:
- change get_all_items_in_game just to iterate over the hash table rather than go through rooms and inventory 
- will we need to free the item hash table in game if we are already freeing all the items within rooms and inventory?
- in issue 315/316: change add_effect and add_condition to only require game and source item_id (no more room needed)

Side note: updated documentation in game.h to reflect when parameters are pointers.

**Comment by patrickkoopmans** [comment]
> Added hash table of items (all_items) into game struct.
> 
> Added functions into game.h/game.c:
> int add_item_to_game(game_t *game, item_t *item);
> item_t *get_item_from_game(game_t *g, char *item_id);
> 
> Things to consider:
> 
> * change get_all_items_in_game just to iterate over the hash table rather than go through rooms and inventory
> * will we need to free the item hash table in game if we are already freeing all the items within rooms and inventory?
> * in issue 315/316: change add_effect and add_condition to only require game and source item_id (no more room needed) (now done)
> 
> Side note: updated documentation in game.h to reflect when parameters are pointers.

Regarding your things to consider: 
1. I agree with this, will make the changes now
2. I don't think so. 
3. Thank you for changing this.

**Comment by patrickkoopmans** [comment]
> > Added hash table of items (all_items) into game struct.
> > Added functions into game.h/game.c:
> > int add_item_to_game(game_t *game, item_t *item);
> > item_t *get_item_from_game(game_t *g, char *item_id);
> > Things to consider:
> > 
> > * change get_all_items_in_game just to iterate over the hash table rather than go through rooms and inventory
> > * will we need to free the item hash table in game if we are already freeing all the items within rooms and inventory?
> > * in issue 315/316: change add_effect and add_condition to only require game and source item_id (no more room needed) (now done)
> > 
> > Side note: updated documentation in game.h to reflect when parameters are pointers.
> 
> Regarding your things to consider:
> 
> 1. I agree with this, will make the changes now (EDIT: This has been done)
> 2. I don't think so.
> 3. Thank you for changing this.

**Comment by jessxia** [comment]
Fixed a small variable name issue within get_all_items_in_game(). Now compiles yay

**Comment by jessxia** [comment]
Closing statement:
In this issue, we added an item hash table into the game struct, which stores the pointers to all items in the game using their unique item_ids as the hash keys. This will not be a problem since WDL has in their specifications that all items must have unique ids. 

We added the following functions: 

- int add_item_to_game(game_t *game, item_t *item), which adds an item to a game hash table
- item_t *get_item_from_game(game_t *g, char *item_id), which returns the item matching the item_id given from the game's item hash table

We also changed get_all_items_in_game to iterate over the hash table of items in the game rather than pass through the item hash table in every room and the inventory since there is now an easier way to access all items.
